File: src/memory_trust.py
"""Trust-by-provenance classifier (memory-trust metaplan T1/T3/T7).

One function decides, for both hosts, whether a memory carries force at
injection or stays behind the untrusted firewall:

- hand-authored (source_type == "human"): ALWAYS trusted — no toggle.
- explicitly pinned: ALWAYS trusted — a pin is a user endorsement (T3).
- anything else (ai / procedural / auto_extracted): trusted only when
  the MASTER toggle is on AND the record's kind switch is on (T7).
- kind "raw" (and anything unrecognized) is never auto-trusted.

The classifier keys on the record's own fields, never on digest-array
membership: fm's digest "pinned" array auto-includes every persona-kind
record, so being listed there is NOT an endorsement (audit finding F7).
"""

from typing import Any, Dict, Mapping, Tuple

BEHAVIOR_KINDS = ("instruction", "persona")
KNOWLEDGE_KINDS = ("fact", "episodic", "fabric", "wiki")
TRUSTABLE_KINDS = BEHAVIOR_KINDS + KNOWLEDGE_KINDS

# e's T7 defaults: flipping the master trusts auto-captured knowledge;
# auto-captured behavior needs a second, deliberate flip.
DEFAULT_KIND_TRUST: Dict[str, bool] = {
    "instruction": False,
    "persona": False,
    "fact": True,
    "episodic": True,
    "fabric": True,
    "wiki": True,
}

MASTER_PREF = "memory_trust_auto"
KINDS_PREF = "memory_trust_auto_kinds"


def trust_prefs(prefs: Any) -> Tuple[bool, Dict[str, bool]]:
    """Sanitize raw per-user prefs into (master, kind_switches).

    Unknown kind keys are dropped, values coerced to bool, missing kinds
    take the T7 defaults. Malformed prefs read as all-off master.
    """
    if not isinstance(prefs, Mapping):
        return False, dict(DEFAULT_KIND_TRUST)
    master = bool(prefs.get(MASTER_PREF, False))
    kinds = dict(DEFAULT_KIND_TRUST)
    raw_kinds = prefs.get(KINDS_PREF)
    if isinstance(raw_kinds, Mapping):
        for kind, value in raw_kinds.items():
            if kind in kinds:
                kinds[kind] = bool(value)
    return master, kinds


def _field(entry: Any, name: str, default: Any = None) -> Any:
    if isinstance(entry, Mapping):
        return entry.get(name, default)
    return getattr(entry, name, default)


def trusted(entry: Any, prefs: Any) -> bool:
    """True when this memory carries force at injection.

    `entry` is anything record-shaped: a MemoryRecord, an enriched digest
    pinned entry, or a plain dict off the fm wire. Requires explicit
    source_type/pinned fields — absent fields read as untrusted, so a
    degraded payload fails CLOSED.
    """
    if str(_field(entry, "source_type") or "") == "human":
        return True
    if bool(_field(entry, "pinned", False)):
        return True
    kind = str(_field(entry, "kind") or "")
    if kind not in TRUSTABLE_KINDS:
        return False
    master, kinds = trust_prefs(prefs)
    return master and kinds.get(kind, False)

File: static/js/util/memoryTrust.js
// Memory trust + signal presentation helpers (memory-trust metaplan
// T2/T5/T7). Pure functions — mirrored from src/memory_trust.py; keep the
// two in lockstep. Node-testable (no DOM).

export const BEHAVIOR_KINDS = ['instruction', 'persona'];
export const KNOWLEDGE_KINDS = ['fact', 'episodic', 'fabric', 'wiki'];
export const TRUSTABLE_KINDS = [...BEHAVIOR_KINDS, ...KNOWLEDGE_KINDS];

// T7 defaults: flipping the master trusts auto-captured knowledge;
// auto-captured behavior needs a second, deliberate flip.
export const DEFAULT_KIND_TRUST = {
  instruction: false,
  persona: false,
  fact: true,
  episodic: true,
  fabric: true,
  wiki: true,
};

export function trustPrefs(prefs) {
  const safe = prefs && typeof prefs === 'object' ? prefs : {};
  const master = Boolean(safe.memory_trust_auto);
  const kinds = { ...DEFAULT_KIND_TRUST };
  const raw = safe.memory_trust_auto_kinds;
  if (raw && typeof raw === 'object') {
    for (const [kind, value] of Object.entries(raw)) {
      if (kind in kinds) kinds[kind] = Boolean(value);
    }
  }
  return { master, kinds };
}

// Mirror of src/memory_trust.trusted(): human always; explicit pin always;
// otherwise master AND kind switch; raw/unknown never. Absent fields fail
// closed.
export function isTrusted(record, prefs) {
  if (!record || typeof record !== 'object') return false;
  if (String(record.source_type || '') === 'human') return true;
  if (record.pinned) return true;
  const kind = String(record.kind || '');
  if (!TRUSTABLE_KINDS.includes(kind)) return false;
  const { master, kinds } = trustPrefs(prefs);
  return master && Boolean(kinds[kind]);
}

// T5: bucketed chip text, raw float on hover (caller puts raw in title).
export function scoreBucket(value) {
  const v = Number(value);
  if (!Number.isFinite(v)) return null;
  if (v < 0.34) return 'low';
  if (v < 0.67) return 'med';
  return 'high';
}

const PROVENANCE_LABEL = {
  human: 'human',
  ai: 'ai',
  auto_extracted: 'auto',
  procedural: 'procedural',
};

// Chip data for one memory card: [{label, cls, title}] — presentation
// only, no DOM. Chips appear only when they carry information (defaults
// and empty fields render nothing).
export function memoryChips(record, prefs) {
  if (!record || typeof record !== 'object') return [];
  const chips = [];

  chips.push(
    isTrusted(record, prefs)
      ? { label: 'trusted', cls: 'trusted', title: 'Carries force at injection (endorsed by you or by your trust toggles)' }
      : { label: 'reference', cls: 'reference', title: 'Shown to the AI as untrusted reference data only' },
  );

  const kind = String(record.kind || '');
  if (kind && kind !== String(record.category || '')) {
    chips.push({ label: kind, cls: 'kind', title: `Engine kind: ${kind}` });
  }

  const provenance = PROVENANCE_LABEL[String(record.source_type || '')];
  if (provenance) {
    chips.push({ label: provenance, cls: 'provenance', title: `Captured by: ${record.source_type}${record.source ? ` (via ${record.source})` : ''}` });
  }

  const workspace = String(record.workspace_id || '');
  if (workspace && workspace !== 'global') {
    chips.push({ label: workspace, cls: 'workspace', title: `Scoped to workspace ${workspace}` });
  }

  if (record.archived) chips.push({ label: 'archived', cls: 'archived', title: 'Archived — excluded from recall' });
  if (record.exempt_from_decay) chips.push({ label: 'no-decay', cls: 'exempt', title: 'Exempt from decay' });
  if (record.exempt_from_dedup) chips.push({ label: 'no-dedup', cls: 'exempt', title: 'Exempt from dedup' });

  for (const [key, short] of [['trust_score', 'T'], ['confidence_score', 'C'], ['importance_score', 'I']]) {
    const bucket = scoreBucket(record[key]);
    if (bucket !== null) {
      chips.push({
        label: `${short}:${bucket}`,
        cls: `score score-${bucket}`,
        title: `${key.replace('_', ' ')}: ${Number(record[key]).toFixed(3)}`,
      });
    }
  }

  return chips;
}
